In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
sns.set(color_codes=True)
import warnings
warnings.filterwarnings('ignore')
data=pd.read_csv('Twitter.data',header=None) #Reading the data file.

### Data Exploration

In [2]:
data.head()

0   1   2   3   4   5   6   7   8   9  ...    68   69  70  71  72  73  74  \
0   0   2   0   0   1   1   1   0   1   0 ...   1.0  1.0   0   2   0   0   1   
1   2   1   0   0   0   0   4   2   1   0 ...   0.0  1.0   2   1   0   0   0   
2   1   0   0   0   0   4   1   1   0   0 ...   1.0  1.0   1   0   0   0   0   
3   1   0   0   1   0   0   1   1   0   0 ...   0.0  1.0   1   0   0   1   0   
4   0   1   0   0   1   2   3   0   1   0 ...   1.0  1.0   0   1   0   0   1   

   75  76   77  
0   1   1  0.0  
1   0   4  0.5  
2   4   1  0.0  
3   0   1  2.5  
4   2   3  0.5  

[5 rows x 78 columns]

In [3]:
colnames = ["NCD","AI","AS(NA)","BL","NAC","AS(NAC)","CS","AT","NA","ADL","NAD","Buzz_Magnitude"]

def colrange(n):
    if n in range(0,7):
        return(colnames[0]+'_'+str(n-0))
    elif n in range(7,14) :
        return(colnames[1]+'_'+str(n-7))
    elif n in range(14,21) :
        return(colnames[2]+'_'+str(n-14))
    elif n in range(21,28) :
        return(colnames[3]+'_'+str(n-21))
    elif n in range(28,35) :
        return(colnames[4]+'_'+str(n-28))
    elif n in range(35,42) :
        return(colnames[5]+'_'+str(n-35))
    elif n in range(42,49) :
        return(colnames[6]+'_'+str(n-42))
    elif n in range(49,56) :
        return(colnames[7]+'_'+str(n-49))
    elif n in range(56,63) :
        return(colnames[8]+'_'+str(n-56))
    elif n in range(63,70) :
        return(colnames[9]+'_'+str(n-63))
    elif n in range(70,77) :
        return(colnames[10]+'_'+str(n-70))
    elif n==77 :
        return(colnames[11])    

data.columns=[colrange(col) for col in data.columns]

In [4]:
data.shape

(583250, 78)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 583250 entries, 0 to 583249
Data columns (total 78 columns):
NCD_0             583250 non-null int64
NCD_1             583250 non-null int64
NCD_2             583250 non-null int64
NCD_3             583250 non-null int64
NCD_4             583250 non-null int64
NCD_5             583250 non-null int64
NCD_6             583250 non-null int64
AI_0              583250 non-null int64
AI_1              583250 non-null int64
AI_2              583250 non-null int64
AI_3              583250 non-null int64
AI_4              583250 non-null int64
AI_5              583250 non-null int64
AI_6              583250 non-null int64
AS(NA)_0          583250 non-null float64
AS(NA)_1          583250 non-null float64
AS(NA)_2          583250 non-null float64
AS(NA)_3          583250 non-null float64
AS(NA)_4          583250 non-null float64
AS(NA)_5          583250 non-null float64
AS(NA)_6          583250 non-null float64
BL_0              583250 non-null f

In [7]:
data.describe()

NCD_0          NCD_1          NCD_2          NCD_3  \
count  583250.000000  583250.000000  583250.000000  583250.000000   
mean      140.339640     136.770147     159.679271     181.592091   
std       431.772639     432.305129     502.057428     574.883713   
min         0.000000       0.000000       0.000000       0.000000   
25%         3.000000       3.000000       4.000000       4.000000   
50%        18.000000      17.000000      21.000000      24.000000   
75%       104.000000     100.000000     115.000000     131.000000   
max     24210.000000   29574.000000   37505.000000   72366.000000   

               NCD_4          NCD_5          NCD_6           AI_0  \
count  583250.000000  583250.000000  583250.000000  583250.000000   
mean      201.097445     220.175371     219.388214      71.038051   
std       630.448432     669.205930     672.182204     196.876718   
min         0.000000       0.000000       0.000000       0.000000   
25%         5.000000       6.000000       6.000000       2.000000   
50%        27.000000      31.000000      30.000000      11.000000   
75%       147.000000     166.000000     164.000000      59.000000   
max     79079.000000   79079.000000   79079.000000   18654.000000   

                AI_1           AI_2       ...                ADL_5  \
count  583250.000000  583250.000000       ...        583250.000000   
mean       69.829631      82.198203       ...             1.136688   
std       202.199758     239.523042       ...             1.432327   
min         0.000000       0.000000       ...             0.000000   
25%         2.000000       2.000000       ...             1.000000   
50%        11.000000      13.000000       ...             1.000000   
75%        57.000000      65.000000       ...             1.090909   
max     22035.000000   29402.000000       ...           262.000000   

               ADL_6          NAD_0          NAD_1          NAD_2  \
count  583250.000000  583250.000000  583250.000000  583250.000000   
mean        1.140372     140.789860     137.181270     160.105922   
std         1.552313     432.624954     433.026611     502.774408   
min         0.000000       0.000000       0.000000       0.000000   
25%         1.000000       3.000000       3.000000       4.000000   
50%         1.000000      18.000000      17.000000      21.000000   
75%         1.091296     104.000000     101.000000     115.000000   
max       295.000000   24301.000000   29574.000000   37505.000000   

               NAD_3          NAD_4          NAD_5          NAD_6  \
count  583250.000000  583250.000000  583250.000000  583250.000000   
mean      182.057440     201.596482     220.705900     219.936864   
std       575.658022     631.258318     670.050977     673.032541   
min         0.000000       0.000000       0.000000       0.000000   
25%         4.000000       5.000000       6.000000       6.000000   
50%        24.000000      27.000000      31.000000      31.000000   
75%       131.000000     148.000000     167.000000     165.000000   
max     72366.000000   79083.000000   79083.000000   79083.000000   

       Buzz_Magnitude  
count   583250.000000  
mean       191.279493  
std        612.352354  
min          0.000000  
25%          4.500000  
50%         25.500000  
75%        139.000000  
max      75724.500000  

[8 rows x 78 columns]

### Checking For Data Leakage

In [8]:
co_matrix=data.corr()
co_matrix["Buzz_Magnitude"].sort_values(ascending=False)

Buzz_Magnitude    1.000000
NCD_6             0.955330
NAD_6             0.955299
NAC_6             0.951809
NCD_5             0.918565
NAD_5             0.918533
NAC_5             0.915110
NA_6              0.905061
NCD_1             0.889884
NAD_1             0.889744
NCD_4             0.886262
NAD_4             0.886251
NAC_1             0.883589
NAC_4             0.883327
NCD_0             0.883316
NAD_0             0.883073
NAC_0             0.875533
NCD_2             0.875168
NAD_2             0.875099
NCD_3             0.873060
NAD_3             0.873026
NAC_2             0.869843
NAC_3             0.869301
NA_5              0.868421
AS(NAC)_6         0.866314
AS(NAC)_0         0.853992
NA_1              0.850840
AS(NAC)_1         0.850498
AS(NAC)_5         0.845093
NA_0              0.843521
                    ...   
AI_3              0.756556
AI_4              0.756219
BL_2              0.091175
CS_2              0.089560
BL_0              0.085147
BL_3              0.085041
C

### Split Data

In [6]:
XFull = data.iloc[:, :-1]
YFull = data.iloc[:, -1]

In [7]:
from sklearn.model_selection import train_test_split
_, sample_data, _, sample_target = train_test_split(XFull, YFull, shuffle = True, test_size = 0.1)

In [8]:
X_train_org, X_test_org, Y_train, Y_test = train_test_split(sample_data[:1000],sample_target[:1000])

In [14]:
def Skewness(data):
    s = (data.mean(),data.median())
    s = pd.concat(s,axis=1)
    s.columns = ['mean','median']
    return(s)
Skewness(data)

mean     median
NCD_0           140.339640  18.000000
NCD_1           136.770147  17.000000
NCD_2           159.679271  21.000000
NCD_3           181.592091  24.000000
NCD_4           201.097445  27.000000
NCD_5           220.175371  31.000000
NCD_6           219.388214  30.000000
AI_0             71.038051  11.000000
AI_1             69.829631  11.000000
AI_2             82.198203  13.000000
AI_3             93.774819  14.000000
AI_4            103.895695  16.000000
AI_5            113.611141  19.000000
AI_6            112.940921  18.000000
AS(NA)_0          0.000202   0.000026
AS(NA)_1          0.000196   0.000025
AS(NA)_2          0.000221   0.000029
AS(NA)_3          0.000242   0.000033
AS(NA)_4          0.000261   0.000037
AS(NA)_5          0.000282   0.000042
AS(NA)_6          0.000279   0.000041
BL_0              0.925390   1.000000
BL_1              0.918146   1.000000
BL_2              0.907052   1.000000
BL_3              0.919954   1.000000
BL_4              0.930657   1.000000
BL_5              0.956070   1.000000
BL_6              0.955341   1.000000
NAC_0           150.662500  20.000000
NAC_1           146.641742  19.000000
...                    ...        ...
CS_6              0.960626   1.000000
AT_0              1.045500   1.000000
AT_1              1.035194   1.000000
AT_2              1.017044   1.000000
AT_3              1.032634   1.000000
AT_4              1.045863   1.000000
AT_5              1.078122   1.000000
AT_6              1.081759   1.000000
NA_0            123.063726  17.000000
NA_1            119.579506  17.000000
NA_2            138.958520  20.000000
NA_3            157.423923  22.000000
NA_4            173.702114  25.000000
NA_5            189.521478  29.000000
NA_6            188.410279  28.000000
ADL_0             1.095255   1.000000
ADL_1             1.083723   1.000000
ADL_2             1.066749   1.000000
ADL_3             1.085214   1.000000
ADL_4             1.100215   1.000000
ADL_5             1.136688   1.000000
ADL_6             1.140372   1.000000
NAD_0           140.789860  18.000000
NAD_1           137.181270  17.000000
NAD_2           160.105922  21.000000
NAD_3           182.057440  24.000000
NAD_4           201.596482  27.000000
NAD_5           220.705900  31.000000
NAD_6           219.936864  31.000000
Buzz_Magnitude  191.279493  25.500000

[78 rows x 2 columns]

#### We will proceed to use MinMax Scaler as independent variables are not normally distributed. Min Max scalar Works well with our data set as compared to Standard Scaler

In [9]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train_org)
X_test = scaler.transform(X_test_org)

## 1 :  Linear regression

In [17]:
from sklearn.linear_model import LinearRegression
lreg = LinearRegression()
from sklearn.model_selection import cross_val_score

In [18]:
CvScores = cross_val_score(lreg, X_train, Y_train, cv=5)
print("CV Scores: {}".format(CvScores))
CvScores

CV Scores: [ 0.91734389  0.79901594  0.94631502 -2.50757904  0.85919064]


array([ 0.91734389,  0.79901594,  0.94631502, -2.50757904,  0.85919064])

In [19]:
Train_Score = CvScores.mean()
print("Train Set Score: {:.2f}".format(Train_Score))
lreg.fit(X_train, Y_train)
Test_Score = lreg.score(X_test,Y_test)
print("Test Set Score: {:.2f}".format(Test_Score))

Train Set Score: 0.20
Test Set Score: 0.96


### Bagging

In [20]:
from sklearn.ensemble import BaggingRegressor
from sklearn.model_selection import GridSearchCV

bag_lreg = BaggingRegressor(lreg, bootstrap = True,oob_score=True,random_state=0)
param_grid = {'n_estimators': [50,100,200]}
grid_bag_lreg = GridSearchCV(bag_lreg, param_grid, cv = 5,return_train_score=True)
grid_bag_lreg.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=BaggingRegressor(base_estimator=LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False),
         bootstrap=True, bootstrap_features=False, max_features=1.0,
         max_samples=1.0, n_estimators=10, n_jobs=None, oob_score=True,
         random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [21]:
print("Best parameters: {}".format(grid_bag_lreg.best_params_))

Best parameters: {'n_estimators': 50}


In [22]:
bag_lreg = BaggingRegressor(lreg,n_estimators = grid_bag_lreg.best_params_['n_estimators'],bootstrap = True,oob_score=True,random_state=0)
bag_lreg.fit(X_train, Y_train)
print("Best cross-validation accuracy: {:.2f}".format(grid_bag_lreg.best_score_))
print("Train Set Score: {}".format(grid_bag_lreg.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_bag_lreg.score(X_test,Y_test)))
print("Oob Score: {}".format(bag_lreg.oob_score_))

Best cross-validation accuracy: 0.62
Train Set Score: 0.9360635314602892
Test Set Score: 0.9663858785171237
Oob Score: 0.7584193652369599


### Pasting

In [23]:
p_lreg = BaggingRegressor(lreg, bootstrap = False,random_state=0)
param_grid = {'n_estimators': [50,100,200]}
grid_p_lreg = GridSearchCV(p_lreg, param_grid, cv = 5,return_train_score=True)
grid_p_lreg.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=BaggingRegressor(base_estimator=LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False),
         bootstrap=False, bootstrap_features=False, max_features=1.0,
         max_samples=1.0, n_estimators=10, n_jobs=None, oob_score=False,
         random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [24]:
print("Best parameters: {}".format(grid_p_lreg.best_params_))
print("Best cross-validation accuracy: {:.2f}".format(grid_bag_lreg.best_score_))
print("Train Set Score: {}".format(grid_bag_lreg.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_bag_lreg.score(X_test,Y_test)))

Best parameters: {'n_estimators': 100}
Best cross-validation accuracy: 0.62
Train Set Score: 0.9360635314602892
Test Set Score: 0.9663858785171237


### AdaBoost

In [25]:
from sklearn.ensemble import AdaBoostRegressor

adbst_lreg = AdaBoostRegressor(lreg,random_state=0)
param_grid = {'learning_rate': [0.01,0.03,0.1,0.3,1.0],
              'n_estimators': [50,100,200]}
grid_adbst_lreg = GridSearchCV(adbst_lreg, param_grid, cv = 5,return_train_score=True)
grid_adbst_lreg.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=AdaBoostRegressor(base_estimator=LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False),
         learning_rate=1.0, loss='linear', n_estimators=50, random_state=0),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'learning_rate': [0.01, 0.03, 0.1, 0.3, 1.0], 'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [26]:
print("Best cross-validation accuracy: {:.2f}".format(grid_adbst_lreg.best_score_))
print("Best parameters: {}".format(grid_adbst_lreg.best_params_))
print("Train Set Score: {}".format(grid_adbst_lreg.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_adbst_lreg.score(X_test,Y_test)))

Best cross-validation accuracy: 0.41
Best parameters: {'learning_rate': 0.03, 'n_estimators': 50}
Train Set Score: 0.9666407538206327
Test Set Score: 0.9520733731182087


### Gradient Boosting

In [27]:
from  sklearn.ensemble import GradientBoostingRegressor

gbrt_lreg = GradientBoostingRegressor(random_state=0)
param_grid = {'max_depth':[1,2],
              'learning_rate' : [0.01,0.03,0.1,0.3,1.0],
              'n_estimators': [50,100,200]}
grid_gbrt_lreg = GridSearchCV(gbrt_lreg, param_grid, cv = 5,return_train_score=True)
grid_gbrt_lreg.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=3, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_sampl...ate=0, subsample=1.0, tol=0.0001,
             validation_fraction=0.1, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'max_depth': [1, 2], 'learning_rate': [0.01, 0.03, 0.1, 0.3, 1.0], 'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [28]:
print("Best cross-validation accuracy: {:.2f}".format(grid_gbrt_lreg.best_score_))
print("Best parameters: {}".format(grid_gbrt_lreg.best_params_))
print("Train Set Score: {}".format(grid_gbrt_lreg.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_gbrt_lreg.score(X_test,Y_test)))

Best cross-validation accuracy: 0.91
Best parameters: {'learning_rate': 0.3, 'max_depth': 1, 'n_estimators': 50}
Train Set Score: 0.9643050053938437
Test Set Score: 0.7469108841142499


## 2 : KNN Regressor 

In [29]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import GridSearchCV
model_knn = KNeighborsRegressor()

In [30]:
param_knn = {'n_neighbors':[1, 5, 10, 15, 20]}
print("Defined Parameters:\n{}".format(param_knn))

Defined Parameters:
{'n_neighbors': [1, 5, 10, 15, 20]}


In [31]:
grid_knn = GridSearchCV(model_knn, param_grid = param_knn, cv=5, return_train_score=True)
grid_knn.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
          metric_params=None, n_jobs=None, n_neighbors=5, p=2,
          weights='uniform'),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_neighbors': [1, 5, 10, 15, 20]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [32]:
print("Best cross-validation accuracy: {:.2f}".format(grid_knn.best_score_))
print("Best parameters: {}".format(grid_knn.best_params_))
print("Train Set Score: {}".format(grid_knn.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_knn.score(X_test,Y_test)))

Best cross-validation accuracy: 0.86
Best parameters: {'n_neighbors': 5}
Train Set Score: 0.9217826281635184
Test Set Score: 0.6218434686971733


In [33]:
best_model_knn = KNeighborsRegressor(n_neighbors = grid_knn.best_params_['n_neighbors'])

### Bagging

In [34]:
bag_knn = BaggingRegressor(best_model_knn, bootstrap = True,oob_score=True,random_state=0)
param_grid = {'n_estimators': [50,100,200]}
grid_bag_knn = GridSearchCV(bag_knn, param_grid, cv = 5,return_train_score=True)
grid_bag_knn.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=BaggingRegressor(base_estimator=KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
          metric_params=None, n_jobs=None, n_neighbors=5, p=2,
          weights='uniform'),
         bootstrap=True, bootstrap_features=False, max_features=1.0,
         max_samples=1.0, n_estimators=10, n_jobs=None, oob_score=True,
         random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [35]:
print("Best parameters: {}".format(grid_bag_knn.best_params_))

Best parameters: {'n_estimators': 200}


In [36]:
best_bag_knn = BaggingRegressor(best_model_knn,n_estimators = grid_bag_knn.best_params_['n_estimators'],bootstrap = True,oob_score=True,random_state=0)
best_bag_knn.fit(X_train, Y_train)
print("Best cross-validation accuracy: {:.2f}".format(grid_bag_knn.best_score_))
print("Train Set Score: {}".format(grid_bag_knn.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_bag_knn.score(X_test,Y_test)))
print("Oob Score: {}".format(best_bag_knn.oob_score_))

Best cross-validation accuracy: 0.86
Train Set Score: 0.9233014753742587
Test Set Score: 0.6176881839957373
Oob Score: 0.8810135647834894


### Pasting

In [37]:
p_knn = BaggingRegressor(best_model_knn, bootstrap = False,random_state=0)
param_grid = {'n_estimators': [50,100,200]}
grid_p_knn = GridSearchCV(p_knn, param_grid, cv = 5,return_train_score=True)
grid_p_knn.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=BaggingRegressor(base_estimator=KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
          metric_params=None, n_jobs=None, n_neighbors=5, p=2,
          weights='uniform'),
         bootstrap=False, bootstrap_features=False, max_features=1.0,
         max_samples=1.0, n_estimators=10, n_jobs=None, oob_score=False,
         random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [38]:
print("Best cross-validation accuracy: {:.2f}".format(grid_p_knn.best_score_))
print("Best parameters: {}".format(grid_p_knn.best_params_))
print("Train Set Score: {}".format(grid_p_knn.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_p_knn.score(X_test,Y_test)))

Best cross-validation accuracy: 0.86
Best parameters: {'n_estimators': 50}
Train Set Score: 0.9217826281635184
Test Set Score: 0.621843468697173


### AdaBoost

In [39]:
adbst_knn = AdaBoostRegressor(best_model_knn,random_state=0)
param_grid = {'learning_rate': [0.01,0.03,0.1,0.3,1.0],
              'n_estimators': [50,100,200]}
grid_adbst_knn = GridSearchCV(adbst_knn, param_grid, cv = 5,return_train_score=True)
grid_adbst_knn.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=AdaBoostRegressor(base_estimator=KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
          metric_params=None, n_jobs=None, n_neighbors=5, p=2,
          weights='uniform'),
         learning_rate=1.0, loss='linear', n_estimators=50, random_state=0),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'learning_rate': [0.01, 0.03, 0.1, 0.3, 1.0], 'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [40]:
print("Best cross-validation accuracy: {:.2f}".format(grid_adbst_knn.best_score_))
print("Best parameters: {}".format(grid_adbst_knn.best_params_))
print("Train Set Score: {}".format(grid_adbst_knn.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_adbst_knn.score(X_test,Y_test)))

Best cross-validation accuracy: 0.87
Best parameters: {'learning_rate': 0.03, 'n_estimators': 100}
Train Set Score: 0.9761442593244664
Test Set Score: 0.6640348528245732


### Gradient Boosting

In [41]:
gbrt_knn = GradientBoostingRegressor(random_state=0)
param_grid = {'max_depth':[1,2],
              'learning_rate' : [0.01,0.03,0.1,0.3,1.0],
              'n_estimators': [50,100,200]}
grid_gbrt_knn = GridSearchCV(gbrt_knn, param_grid, cv = 5,return_train_score=True)
grid_gbrt_knn.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=3, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_sampl...ate=0, subsample=1.0, tol=0.0001,
             validation_fraction=0.1, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'max_depth': [1, 2], 'learning_rate': [0.01, 0.03, 0.1, 0.3, 1.0], 'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [42]:
print("Best cross-validation accuracy: {:.2f}".format(grid_gbrt_knn.best_score_))
print("Best parameters: {}".format(grid_gbrt_knn.best_params_))
print("Train Set Score: {}".format(grid_gbrt_knn.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_gbrt_knn.score(X_test,Y_test)))

Best cross-validation accuracy: 0.91
Best parameters: {'learning_rate': 0.3, 'max_depth': 1, 'n_estimators': 50}
Train Set Score: 0.9643050053938437
Test Set Score: 0.7469108841142499


## 3 : Stochastic Gradient Descent

In [43]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import SGDRegressor
sgd_reg=SGDRegressor()

In [44]:
param_sgd = {'max_iter': [10, 100, 1000],'learning_rate':['optimal'], 'penalty' :['l1','l2'],'random_state':[0]}
print("Defined Parameters:\n{}".format(param_sgd))

Defined Parameters:
{'max_iter': [10, 100, 1000], 'learning_rate': ['optimal'], 'penalty': ['l1', 'l2'], 'random_state': [0]}


In [45]:
grid_sgd = GridSearchCV(estimator = sgd_reg, param_grid = param_sgd, cv=5, return_train_score=True)
grid_sgd.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=SGDRegressor(alpha=0.0001, average=False, early_stopping=False, epsilon=0.1,
       eta0=0.01, fit_intercept=True, l1_ratio=0.15,
       learning_rate='invscaling', loss='squared_loss', max_iter=None,
       n_iter=None, n_iter_no_change=5, penalty='l2', power_t=0.25,
       random_state=None, shuffle=True, tol=None, validation_fraction=0.1,
       verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'max_iter': [10, 100, 1000], 'learning_rate': ['optimal'], 'penalty': ['l1', 'l2'], 'random_state': [0]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [46]:
print("Best cross-validation accuracy: {:.2f}".format(grid_sgd.best_score_))
print("Best parameters: {}".format(grid_sgd.best_params_))
print("Train Set Score: {}".format(grid_sgd.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_sgd.score(X_test, Y_test)))

Best cross-validation accuracy: -711638552861.15
Best parameters: {'learning_rate': 'optimal', 'max_iter': 1000, 'penalty': 'l2', 'random_state': 0}
Train Set Score: -107484545826.84642
Test Set Score: -79888906589.90059


In [47]:
best_sgd_reg=SGDRegressor(learning_rate=grid_sgd.best_params_['learning_rate'],max_iter=grid_sgd.best_params_['max_iter'],penalty=grid_sgd.best_params_['penalty'],random_state=0)

### Bagging

In [48]:
bag_sgd = BaggingRegressor(best_sgd_reg, bootstrap = True,oob_score=True,random_state=0)
param_grid = {'n_estimators': [50,100,200]}
grid_bag_sgd = GridSearchCV(bag_sgd, param_grid, cv = 5,return_train_score=True)
grid_bag_sgd.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=BaggingRegressor(base_estimator=SGDRegressor(alpha=0.0001, average=False, early_stopping=False, epsilon=0.1,
       eta0=0.01, fit_intercept=True, l1_ratio=0.15,
       learning_rate='optimal', loss='squared_loss', max_iter=1000,
       n_iter=None, n_iter_no_change=5, penalty='l2', power_... n_estimators=10, n_jobs=None, oob_score=True,
         random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [49]:
print("Best parameters: {}".format(grid_bag_sgd.best_params_))

Best parameters: {'n_estimators': 200}


In [50]:
best_bag_sgd = BaggingRegressor(best_sgd_reg,n_estimators = grid_bag_sgd.best_params_['n_estimators'],bootstrap = True,oob_score=True,random_state=0)
best_bag_sgd.fit(X_train, Y_train)
print("Best cross-validation accuracy: {:.2f}".format(grid_bag_sgd.best_score_))
print("Train Set Score: {}".format(grid_bag_sgd.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_bag_sgd.score(X_test,Y_test)))
print("Oob Score: {}".format(best_bag_sgd.oob_score_))

Best cross-validation accuracy: -73497614945.74
Train Set Score: -9553233893.248554
Test Set Score: -3586255036.95716
Oob Score: -52910347519.02106


### Pasting

In [51]:
p_sgd = BaggingRegressor(best_sgd_reg, bootstrap = False,random_state=0)
param_grid = {'n_estimators': [50,100,200]}
grid_p_sgd = GridSearchCV(p_sgd, param_grid, cv = 5,return_train_score=True)
grid_p_sgd.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=BaggingRegressor(base_estimator=SGDRegressor(alpha=0.0001, average=False, early_stopping=False, epsilon=0.1,
       eta0=0.01, fit_intercept=True, l1_ratio=0.15,
       learning_rate='optimal', loss='squared_loss', max_iter=1000,
       n_iter=None, n_iter_no_change=5, penalty='l2', power_...n_estimators=10, n_jobs=None, oob_score=False,
         random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [52]:
print("Best cross-validation accuracy: {:.2f}".format(grid_p_sgd.best_score_))
print("Best parameters: {}".format(grid_p_sgd.best_params_))
print("Train Set Score: {}".format(grid_p_sgd.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_p_sgd.score(X_test,Y_test)))

Best cross-validation accuracy: -8927036776.59
Best parameters: {'n_estimators': 200}
Train Set Score: -1058890255.3373134
Test Set Score: -3307987981.7685046


### AdaBoost

In [53]:
adbst_sgd = AdaBoostRegressor(best_sgd_reg,random_state=0)
param_grid = {'learning_rate': [0.01,0.03,0.1,0.3,1.0],
              'n_estimators': [50,100,200]}
grid_adbst_sgd = GridSearchCV(adbst_sgd, param_grid, cv = 5,return_train_score=True)
grid_adbst_sgd.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=AdaBoostRegressor(base_estimator=SGDRegressor(alpha=0.0001, average=False, early_stopping=False, epsilon=0.1,
       eta0=0.01, fit_intercept=True, l1_ratio=0.15,
       learning_rate='optimal', loss='squared_loss', max_iter=1000,
       n_iter=None, n_iter_no_change=5, penalty='l2', power_t=0.25,
       random_state=0, shuffle=True, tol=None, validation_fraction=0.1,
       verbose=0, warm_start=False),
         learning_rate=1.0, loss='linear', n_estimators=50, random_state=0),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'learning_rate': [0.01, 0.03, 0.1, 0.3, 1.0], 'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [54]:
print("Best cross-validation accuracy: {:.2f}".format(grid_adbst_sgd.best_score_))
print("Best parameters: {}".format(grid_adbst_sgd.best_params_))
print("Train Set Score: {}".format(grid_adbst_sgd.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_adbst_sgd.score(X_test,Y_test)))

Best cross-validation accuracy: -5485318331.42
Best parameters: {'learning_rate': 0.03, 'n_estimators': 200}
Train Set Score: -925668160.083588
Test Set Score: -1690038363.2698963


### Gradient Boosting

In [55]:
gbrt_sgd = GradientBoostingRegressor(random_state=0)
param_grid = {'max_depth':[1,2],
              'learning_rate' : [0.01,0.03,0.1,0.3,1.0],
              'n_estimators': [50,100,200]}
grid_gbrt_sgd = GridSearchCV(gbrt_sgd, param_grid, cv = 5,return_train_score=True)
grid_gbrt_sgd.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=3, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_sampl...ate=0, subsample=1.0, tol=0.0001,
             validation_fraction=0.1, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'max_depth': [1, 2], 'learning_rate': [0.01, 0.03, 0.1, 0.3, 1.0], 'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [56]:
print("Best cross-validation accuracy: {:.2f}".format(grid_gbrt_sgd.best_score_))
print("Best parameters: {}".format(grid_gbrt_sgd.best_params_))
print("Train Set Score: {}".format(grid_gbrt_sgd.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_gbrt_sgd.score(X_test,Y_test)))

Best cross-validation accuracy: 0.91
Best parameters: {'learning_rate': 0.3, 'max_depth': 1, 'n_estimators': 50}
Train Set Score: 0.9643050053938437
Test Set Score: 0.7469108841142499


## 4 : Support Vector Machines

In [57]:
from sklearn.svm import SVR
model_RadSVM = SVR(kernel = 'rbf')
param_RadSVM = {'C': [0.001, 0.01, 0.1, 1, 10, 100],'gamma': [0.001, 0.01, 0.1, 1, 10, 100]}
print("Defined Parameters:\n{}".format(param_RadSVM))

Defined Parameters:
{'C': [0.001, 0.01, 0.1, 1, 10, 100], 'gamma': [0.001, 0.01, 0.1, 1, 10, 100]}


In [58]:
grid_RadSVM = GridSearchCV(model_RadSVM, param_grid = param_RadSVM, cv=5, return_train_score=True)
grid_RadSVM.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1,
  gamma='auto_deprecated', kernel='rbf', max_iter=-1, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'C': [0.001, 0.01, 0.1, 1, 10, 100], 'gamma': [0.001, 0.01, 0.1, 1, 10, 100]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [59]:
print("Best parameters: {}".format(grid_RadSVM.best_params_))
grid_RadSVMdata = pd.DataFrame(grid_RadSVM.cv_results_)
grid_RadSVMdata = grid_RadSVMdata[['params','mean_train_score','mean_test_score']]
grid_RadSVMdata

Best parameters: {'C': 100, 'gamma': 0.1}


params  mean_train_score  mean_test_score
0   {'C': 0.001, 'gamma': 0.001}         -0.146889        -0.155298
1    {'C': 0.001, 'gamma': 0.01}         -0.146881        -0.155289
2     {'C': 0.001, 'gamma': 0.1}         -0.146852        -0.155258
3       {'C': 0.001, 'gamma': 1}         -0.146866        -0.155273
4      {'C': 0.001, 'gamma': 10}         -0.146821        -0.155226
5     {'C': 0.001, 'gamma': 100}         -0.146799        -0.155198
6    {'C': 0.01, 'gamma': 0.001}         -0.146879        -0.155288
7     {'C': 0.01, 'gamma': 0.01}         -0.146792        -0.155200
8      {'C': 0.01, 'gamma': 0.1}         -0.146505        -0.154889
9        {'C': 0.01, 'gamma': 1}         -0.146649        -0.155036
10      {'C': 0.01, 'gamma': 10}         -0.146192        -0.154567
11     {'C': 0.01, 'gamma': 100}         -0.146029        -0.154418
12    {'C': 0.1, 'gamma': 0.001}         -0.146779        -0.155189
13     {'C': 0.1, 'gamma': 0.01}         -0.145904        -0.154314
14      {'C': 0.1, 'gamma': 0.1}         -0.142836        -0.150991
15        {'C': 0.1, 'gamma': 1}         -0.144002        -0.152162
16       {'C': 0.1, 'gamma': 10}         -0.141010        -0.149256
17      {'C': 0.1, 'gamma': 100}         -0.141626        -0.149870
18      {'C': 1, 'gamma': 0.001}         -0.145773        -0.154196
19       {'C': 1, 'gamma': 0.01}         -0.136877        -0.145287
20        {'C': 1, 'gamma': 0.1}         -0.104351        -0.110242
21          {'C': 1, 'gamma': 1}         -0.106066        -0.111492
22         {'C': 1, 'gamma': 10}         -0.107636        -0.115202
23        {'C': 1, 'gamma': 100}         -0.112700        -0.121562
24     {'C': 10, 'gamma': 0.001}         -0.135581        -0.144117
25      {'C': 10, 'gamma': 0.01}         -0.046188        -0.054849
26       {'C': 10, 'gamma': 0.1}          0.267427         0.275474
27         {'C': 10, 'gamma': 1}          0.126601         0.131520
28        {'C': 10, 'gamma': 10}         -0.044923        -0.061840
29       {'C': 10, 'gamma': 100}         -0.066791        -0.090868
30    {'C': 100, 'gamma': 0.001}         -0.035588        -0.045281
31     {'C': 100, 'gamma': 0.01}          0.553009         0.542266
32      {'C': 100, 'gamma': 0.1}          0.761143         0.758410
33        {'C': 100, 'gamma': 1}          0.424279         0.390953
34       {'C': 100, 'gamma': 10}          0.229099         0.118080
35      {'C': 100, 'gamma': 100}          0.192894         0.044437

In [60]:
print("Best cross-validation accuracy: {:.2f}".format(grid_RadSVM.best_score_))
print("Best parameters: {}".format(grid_RadSVM.best_params_))
print("Train Set Score: {}".format(grid_RadSVM.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_RadSVM.score(X_test, Y_test)))

Best cross-validation accuracy: 0.76
Best parameters: {'C': 100, 'gamma': 0.1}
Train Set Score: 0.7850569855945213
Test Set Score: 0.2657644395494879


In [61]:
best_model_RadSVM=SVR(kernel = 'rbf',C=grid_RadSVM.best_params_['C'],gamma=grid_RadSVM.best_params_['gamma'])

### Bagging

In [62]:
bag_RadSVM = BaggingRegressor(best_model_RadSVM, bootstrap = True,oob_score=True,random_state=0)
param_grid = {'n_estimators': [50,100,200]}
grid_bag_RadSVM = GridSearchCV(bag_RadSVM, param_grid, cv = 5,return_train_score=True)
grid_bag_RadSVM.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=BaggingRegressor(base_estimator=SVR(C=100, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma=0.1,
  kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False),
         bootstrap=True, bootstrap_features=False, max_features=1.0,
         max_samples=1.0, n_estimators=10, n_jobs=None, oob_score=True,
         random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [63]:
print("Best parameters: {}".format(grid_bag_RadSVM.best_params_))

Best parameters: {'n_estimators': 100}


In [64]:
best_bag_RadSVM = BaggingRegressor(best_model_RadSVM,n_estimators = grid_bag_RadSVM.best_params_['n_estimators'],bootstrap = True,oob_score=True,random_state=0)
best_bag_RadSVM.fit(X_train, Y_train)
print("Best cross-validation accuracy: {:.2f}".format(grid_bag_RadSVM.best_score_))
print("Train Set Score: {}".format(grid_bag_RadSVM.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_bag_RadSVM.score(X_test,Y_test)))
print("Oob Score: {}".format(best_bag_RadSVM.oob_score_))

Best cross-validation accuracy: 0.76
Train Set Score: 0.7809271313912293
Test Set Score: 0.26336218549677937
Oob Score: 0.7612752237174518


### Pasting

In [65]:
p_RadSVM = BaggingRegressor(best_model_RadSVM, bootstrap = False,random_state=0)
param_grid = {'n_estimators': [50,100,200]}
grid_p_RadSVM = GridSearchCV(p_RadSVM, param_grid, cv = 5,return_train_score=True)
grid_p_RadSVM.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=BaggingRegressor(base_estimator=SVR(C=100, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma=0.1,
  kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False),
         bootstrap=False, bootstrap_features=False, max_features=1.0,
         max_samples=1.0, n_estimators=10, n_jobs=None, oob_score=False,
         random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [66]:
print("Best cross-validation accuracy: {:.2f}".format(grid_p_RadSVM.best_score_))
print("Best parameters: {}".format(grid_p_RadSVM.best_params_))
print("Train Set Score: {}".format(grid_p_RadSVM.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_p_RadSVM.score(X_test,Y_test)))

Best cross-validation accuracy: 0.76
Best parameters: {'n_estimators': 50}
Train Set Score: 0.7850569855945213
Test Set Score: 0.2657644395494879


### AdaBoost

In [67]:
adbst_RadSVM = AdaBoostRegressor(best_model_RadSVM,random_state=0)
param_grid = {'learning_rate': [0.01,0.03,0.1,0.3,1.0],
              'n_estimators': [50,100,200]}
grid_adbst_RadSVM = GridSearchCV(adbst_RadSVM, param_grid, cv = 5,return_train_score=True)
grid_adbst_RadSVM.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=AdaBoostRegressor(base_estimator=SVR(C=100, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma=0.1,
  kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False),
         learning_rate=1.0, loss='linear', n_estimators=50, random_state=0),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'learning_rate': [0.01, 0.03, 0.1, 0.3, 1.0], 'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [68]:
print("Best cross-validation accuracy: {:.2f}".format(grid_adbst_RadSVM.best_score_))
print("Best parameters: {}".format(grid_adbst_RadSVM.best_params_))
print("Train Set Score: {}".format(grid_adbst_RadSVM.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_adbst_RadSVM.score(X_test,Y_test)))

Best cross-validation accuracy: 0.88
Best parameters: {'learning_rate': 0.3, 'n_estimators': 50}
Train Set Score: 0.9276826565485606
Test Set Score: 0.42058043658337174


### Gradient Boosting

In [69]:
gbrt_RadSVM = GradientBoostingRegressor(random_state=0)
param_grid = {'max_depth':[1,2],
              'learning_rate' : [0.01,0.03,0.1,0.3,1.0],
              'n_estimators': [50,100,200]}
grid_gbrt_RadSVM = GridSearchCV(gbrt_RadSVM, param_grid, cv = 5,return_train_score=True)
grid_gbrt_RadSVM.fit(X_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=3, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_sampl...ate=0, subsample=1.0, tol=0.0001,
             validation_fraction=0.1, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'max_depth': [1, 2], 'learning_rate': [0.01, 0.03, 0.1, 0.3, 1.0], 'n_estimators': [50, 100, 200]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [70]:
print("Best cross-validation accuracy: {:.2f}".format(grid_gbrt_RadSVM.best_score_))
print("Best parameters: {}".format(grid_gbrt_RadSVM.best_params_))
print("Train Set Score: {}".format(grid_gbrt_RadSVM.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_gbrt_RadSVM.score(X_test,Y_test)))

Best cross-validation accuracy: 0.91
Best parameters: {'learning_rate': 0.3, 'max_depth': 1, 'n_estimators': 50}
Train Set Score: 0.9643050053938437
Test Set Score: 0.7469108841142499


## Principal Components Analysis

In [13]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.95)
X_reduced_train = pca.fit_transform(X_train)
X_reduced_test = pca.transform(X_test)

In [14]:
X_reduced_train

array([[ -2.28970620e-01,  -1.17548199e-02,  -1.21755077e-01, ...,
         -2.73881541e-02,   1.56835513e-02,   7.46708731e-03],
       [  7.40617520e-01,   1.93652192e-01,  -1.25140955e-02, ...,
         -1.61879203e-01,   7.10926300e-01,   2.06954638e-04],
       [ -2.39321565e-01,  -9.61155022e-03,  -9.25995381e-02, ...,
         -3.50237237e-02,   1.66773494e-03,  -2.42314928e-02],
       ..., 
       [ -2.26894351e-01,  -1.23149760e-02,  -5.06825078e-02, ...,
         -4.36226997e-02,  -1.12389330e-03,   4.14938859e-02],
       [ -2.29053024e-01,  -1.44871049e-02,  -1.48203999e-01, ...,
         -3.75252268e-02,   1.39751935e-03,  -2.46395446e-02],
       [ -3.10347564e-01,   3.61042112e-02,   3.92195919e-01, ...,
         -1.48921475e-02,   5.31182977e-03,   1.40779651e-02]])

In [15]:
X_reduced_test

array([[ -2.31559729e-01,  -1.35319379e-02,  -1.36872994e-01, ...,
         -3.75146889e-02,   1.51487978e-03,  -2.39300548e-02],
       [ -2.06997642e-01,  -5.30630761e-03,  -1.76976214e-01, ...,
         -4.94357413e-02,   4.16476304e-03,   1.68076562e-02],
       [ -2.94954835e-01,   1.73602390e-02,   2.12375759e-01, ...,
         -2.09255121e-02,   3.01791927e-03,  -2.16058744e-02],
       ..., 
       [  1.29356326e+00,  -1.29460430e+00,   1.86045708e-01, ...,
          5.96165804e-01,   3.00280091e-01,   8.46431481e-03],
       [  1.86720929e+00,  -4.37738360e-01,   1.36457697e-01, ...,
         -1.77143598e-02,   1.43646412e-01,  -4.97757295e-03],
       [ -2.63998663e-01,   7.85966225e-03,   8.26801852e-02, ...,
         -2.10086106e-02,   1.06658939e-02,  -1.01856847e-03]])

In [16]:
pca.explained_variance_ratio_.sum()

0.9608257827630603

In [17]:
1-pca.explained_variance_ratio_.sum()

0.039174217236939701

### 1 :  Linear regression

In [18]:
from sklearn.linear_model import LinearRegression
lreg = LinearRegression()
from sklearn.model_selection import cross_val_score

In [19]:
CvScores = cross_val_score(lreg, X_reduced_train, Y_train, cv=5)
print("CV Scores: {}".format(CvScores))

CV Scores: [ 0.7173823   0.86887584  0.72321096  0.82341475  0.80897773]


In [20]:
Train_Score = CvScores.mean()
print("Train Set Score: {:.2f}".format(Train_Score))
lreg.fit(X_reduced_train, Y_train)
Test_Score = lreg.score(X_reduced_test,Y_test)
print("Test Set Score: {:.2f}".format(Test_Score))

Train Set Score: 0.79
Test Set Score: 0.72


In [21]:
report_table = [['Linear Regression', 'None', Train_Score, Test_Score]]

### 2 : KNN Regressor 

In [22]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import GridSearchCV
model_knn = KNeighborsRegressor()

In [23]:
param_knn = {'n_neighbors':[1, 5, 10, 15, 20]}
print("Defined Parameters:\n{}".format(param_knn))

Defined Parameters:
{'n_neighbors': [1, 5, 10, 15, 20]}


In [24]:
grid_knn = GridSearchCV(model_knn, param_grid = param_knn, cv=5, return_train_score=True)
grid_knn.fit(X_reduced_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
          metric_params=None, n_jobs=None, n_neighbors=5, p=2,
          weights='uniform'),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_neighbors': [1, 5, 10, 15, 20]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [25]:
print("Best cross-validation accuracy: {:.2f}".format(grid_knn.best_score_))
print("Best parameters: {}".format(grid_knn.best_params_))
print("Train Set Score: {}".format(grid_knn.score(X_reduced_train, Y_train)))
print("Test Set Score: {}".format(grid_knn.score(X_reduced_test,Y_test)))

Best cross-validation accuracy: 0.73
Best parameters: {'n_neighbors': 5}
Train Set Score: 0.7729707402651411
Test Set Score: 0.6634346085568215


In [26]:
report_table = report_table + [['KNN Regression', 'K = 5', grid_knn.score(X_reduced_train, Y_train), grid_knn.score(X_reduced_test,Y_test)]]

### 3 : Ridge

In [27]:
from sklearn.linear_model import Ridge
model_ridge = Ridge()
param_ridge = {'alpha': [0.001,0.003,0.01,0.03,0.1,0.3,1,3,10,30,100,300,1000]}
print("Defined Parameters:\n{}".format(param_ridge))

Defined Parameters:
{'alpha': [0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3, 10, 30, 100, 300, 1000]}


In [28]:
grid_ridge = GridSearchCV(estimator = model_ridge,param_grid = param_ridge, cv=5, return_train_score=True)
grid_ridge.fit(X_reduced_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'alpha': [0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3, 10, 30, 100, 300, 1000]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [29]:
print("Best cross-validation accuracy: {:.2f}".format(grid_ridge.best_score_))
print("Best parameters: {}".format(grid_ridge.best_params_))
print("Train Set Score: {}".format(grid_ridge.score(X_reduced_train, Y_train)))
print("Test Set Score: {}".format(grid_ridge.score(X_reduced_test, Y_test)))

Best cross-validation accuracy: 0.83
Best parameters: {'alpha': 30}
Train Set Score: 0.8329935912876725
Test Set Score: 0.8061452850650876


In [30]:
report_table = report_table + [['Ridge Regression', 'alpha = 0.3', grid_ridge.score(X_reduced_train, Y_train), grid_ridge.score(X_reduced_test,Y_test)]]

### 4 : Lasso

In [31]:
from  sklearn.linear_model import Lasso
model_lasso = Lasso()

In [32]:
param_lasso = {'alpha': [0.001,0.003,0.01,0.03,0.1,0.3,1,3,10,30,100,300,1000]}
print("Defined Parameters:\n{}".format(param_lasso))

Defined Parameters:
{'alpha': [0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3, 10, 30, 100, 300, 1000]}


In [33]:
grid_lasso = GridSearchCV(model_lasso, param_grid = param_lasso, cv=5, return_train_score=True)
grid_lasso.fit(X_reduced_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'alpha': [0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3, 10, 30, 100, 300, 1000]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [34]:
print("Best cross-validation accuracy: {:.2f}".format(grid_lasso.best_score_))
print("Best parameters: {}".format(grid_lasso.best_params_))
print("Train Set Score: {}".format(grid_lasso.score(X_reduced_train, Y_train)))
print("Test Set Score: {}".format(grid_lasso.score(X_reduced_test, Y_test)))

Best cross-validation accuracy: 0.80
Best parameters: {'alpha': 3}
Train Set Score: 0.8751343199605988
Test Set Score: 0.7332974374871495


In [35]:
report_table = report_table + [['Lasso Regression', 'alpha = 1', grid_lasso.score(X_reduced_train, Y_train), grid_lasso.score(X_reduced_test,Y_test)]]

### 5 : Polynomial Regression

In [36]:
from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(degree=2)

In [37]:
Train_Score_List = []
Test_Score_List = []
lreg_poly = LinearRegression()

X_Train_Poly = poly.fit_transform(X_reduced_train)
X_Test_Poly = poly.transform(X_reduced_test)
PolCVScore = cross_val_score(lreg_poly, X_Train_Poly, Y_train, cv=5)
lreg_poly.fit(X_Train_Poly,Y_train)
Train_ScorePoly = lreg_poly.score(X_Test_Poly,Y_test)

Train_ScorePoly = PolCVScore.mean()
print("Train Set Score: {:.2f}".format(PolCVScore.mean()))
print("Test Set Score: {:.2f}".format(Train_ScorePoly))

Train Set Score: -1.13
Test Set Score: -1.13


In [38]:
report_table = report_table + [['Polynomial Regression', 'Degree = 2', PolCVScore.mean(), Train_ScorePoly]]

### 6 : Stochastic Gradient Descent

In [39]:
from sklearn.linear_model import SGDRegressor
sgd_reg=SGDRegressor()

In [40]:
param_sgd = {'max_iter': [10, 100, 1000],'learning_rate':['optimal'], 'penalty' :['l1','l2'],'random_state':[0]}
print("Defined Parameters:\n{}".format(param_sgd))

Defined Parameters:
{'max_iter': [10, 100, 1000], 'learning_rate': ['optimal'], 'penalty': ['l1', 'l2'], 'random_state': [0]}


In [41]:
grid_sgd = GridSearchCV(estimator = sgd_reg, param_grid = param_sgd, cv=5, return_train_score=True)
grid_sgd.fit(X_reduced_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=SGDRegressor(alpha=0.0001, average=False, early_stopping=False, epsilon=0.1,
       eta0=0.01, fit_intercept=True, l1_ratio=0.15,
       learning_rate='invscaling', loss='squared_loss', max_iter=None,
       n_iter=None, n_iter_no_change=5, penalty='l2', power_t=0.25,
       random_state=None, shuffle=True, tol=None, validation_fraction=0.1,
       verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'max_iter': [10, 100, 1000], 'learning_rate': ['optimal'], 'penalty': ['l1', 'l2'], 'random_state': [0]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [42]:
print("Best parameters: {}".format(grid_sgd.best_params_))
grid_sgdata = pd.DataFrame(grid_sgd.cv_results_)
grid_sgdata = grid_sgdata[['params','mean_train_score','mean_test_score']]
grid_sgdata

Best parameters: {'learning_rate': 'optimal', 'max_iter': 1000, 'penalty': 'l2', 'random_state': 0}


params  mean_train_score  \
0  {'learning_rate': 'optimal', 'max_iter': 10, '...     -1.179081e+19   
1  {'learning_rate': 'optimal', 'max_iter': 10, '...     -1.001653e+19   
2  {'learning_rate': 'optimal', 'max_iter': 100, ...      7.639210e-01   
3  {'learning_rate': 'optimal', 'max_iter': 100, ...      7.648809e-01   
4  {'learning_rate': 'optimal', 'max_iter': 1000,...      8.810997e-01   
5  {'learning_rate': 'optimal', 'max_iter': 1000,...      8.811359e-01   

   mean_test_score  
0    -3.054691e+19  
1    -2.238984e+19  
2     6.933970e-01  
3     6.955235e-01  
4     7.531622e-01  
5     7.539381e-01

In [43]:
print("Best cross-validation accuracy: {:.2f}".format(grid_sgd.best_score_))
print("Best parameters: {}".format(grid_sgd.best_params_))
print("Train Set Score: {}".format(grid_sgd.score(X_reduced_train, Y_train)))
print("Test Set Score: {}".format(grid_sgd.score(X_reduced_test, Y_test)))

Best cross-validation accuracy: 0.75
Best parameters: {'learning_rate': 'optimal', 'max_iter': 1000, 'penalty': 'l2', 'random_state': 0}
Train Set Score: 0.875339292501002
Test Set Score: 0.6966473567478744


In [44]:
report_table = report_table + [['SGD Regression', grid_sgd.best_params_, grid_sgd.score(X_reduced_train, Y_train), grid_sgd.score(X_reduced_test, Y_test)]]

### 7 : Linear SVR

In [45]:
from sklearn.svm import LinearSVR
model_LSVM = LinearSVR()
param_LSVM = {'C':[0.001, 0.01, 0.1, 1, 10, 100, 1000]}
print("Defined Parameters:\n{}".format(param_LSVM))

Defined Parameters:
{'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]}


In [46]:
grid_LSVM = GridSearchCV(model_LSVM, param_grid = param_LSVM, cv=5, return_train_score=True)
grid_LSVM.fit(X_reduced_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=LinearSVR(C=1.0, dual=True, epsilon=0.0, fit_intercept=True,
     intercept_scaling=1.0, loss='epsilon_insensitive', max_iter=1000,
     random_state=None, tol=0.0001, verbose=0),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [47]:
print("Best parameters: {}".format(grid_LSVM.best_params_))
grid_LSVMdata = pd.DataFrame(grid_LSVM.cv_results_)
grid_LSVMdata = grid_LSVMdata[['params','mean_train_score','mean_test_score']]
grid_LSVMdata

Best parameters: {'C': 1000}


params  mean_train_score  mean_test_score
0  {'C': 0.001}         -0.086498        -0.152192
1   {'C': 0.01}         -0.082917        -0.145129
2    {'C': 0.1}         -0.065735        -0.115859
3      {'C': 1}          0.027983         0.007725
4     {'C': 10}          0.548050         0.608811
5    {'C': 100}          0.797658         0.842952
6   {'C': 1000}          0.824513         0.859820

In [48]:
print("Best cross-validation accuracy: {:.2f}".format(grid_LSVM.best_score_))
print("Best parameters: {}".format(grid_LSVM.best_params_))
print("Train Set Score: {}".format(grid_LSVM.score(X_reduced_train, Y_train)))
print("Test Set Score: {}".format(grid_LSVM.score(X_reduced_test, Y_test)))

Best cross-validation accuracy: 0.86
Best parameters: {'C': 1000}
Train Set Score: 0.8247458736852817
Test Set Score: 0.8099948635281169


In [49]:
report_table = report_table + [['Linear SVM', grid_LSVM.best_params_, grid_LSVM.score(X_reduced_train, Y_train), grid_LSVM.score(X_reduced_test, Y_test)]]

### 8 : Support Vector Machines

In [50]:
from sklearn.svm import SVR
model_RadSVM = SVR(kernel = 'rbf')
param_RadSVM = {'C': [0.001, 0.01, 0.1, 1, 10, 100],'gamma': [0.001, 0.01, 0.1, 1, 10, 100]}
print("Defined Parameters:\n{}".format(param_RadSVM))

Defined Parameters:
{'C': [0.001, 0.01, 0.1, 1, 10, 100], 'gamma': [0.001, 0.01, 0.1, 1, 10, 100]}


In [51]:
grid_RadSVM = GridSearchCV(model_RadSVM, param_grid = param_RadSVM, cv=5, return_train_score=True)
grid_RadSVM.fit(X_reduced_train, Y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1,
  gamma='auto_deprecated', kernel='rbf', max_iter=-1, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'C': [0.001, 0.01, 0.1, 1, 10, 100], 'gamma': [0.001, 0.01, 0.1, 1, 10, 100]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [52]:
print("Best parameters: {}".format(grid_RadSVM.best_params_))
grid_RadSVMdata = pd.DataFrame(grid_RadSVM.cv_results_)
grid_RadSVMdata = grid_RadSVMdata[['params','mean_train_score','mean_test_score']]
grid_RadSVMdata

Best parameters: {'C': 100, 'gamma': 0.1}


params  mean_train_score  mean_test_score
0   {'C': 0.001, 'gamma': 0.001}         -0.065296        -0.109261
1    {'C': 0.001, 'gamma': 0.01}         -0.065295        -0.109259
2     {'C': 0.001, 'gamma': 0.1}         -0.065291        -0.109247
3       {'C': 0.001, 'gamma': 1}         -0.065284        -0.109224
4      {'C': 0.001, 'gamma': 10}         -0.065276        -0.109215
5     {'C': 0.001, 'gamma': 100}         -0.065258        -0.109189
6    {'C': 0.01, 'gamma': 0.001}         -0.065294        -0.109259
7     {'C': 0.01, 'gamma': 0.01}         -0.065277        -0.109234
8      {'C': 0.01, 'gamma': 0.1}         -0.065244        -0.109112
9        {'C': 0.01, 'gamma': 1}         -0.065284        -0.109030
10      {'C': 0.01, 'gamma': 10}         -0.065131        -0.108808
11     {'C': 0.01, 'gamma': 100}         -0.064982        -0.108564
12    {'C': 0.1, 'gamma': 0.001}         -0.065273        -0.109231
13     {'C': 0.1, 'gamma': 0.01}         -0.065105        -0.108982
14      {'C': 0.1, 'gamma': 0.1}         -0.065032        -0.108112
15        {'C': 0.1, 'gamma': 1}         -0.065063        -0.106679
16       {'C': 0.1, 'gamma': 10}         -0.063666        -0.105000
17      {'C': 0.1, 'gamma': 100}         -0.062231        -0.102861
18      {'C': 1, 'gamma': 0.001}         -0.065064        -0.108952
19       {'C': 1, 'gamma': 0.01}         -0.063670        -0.106865
20        {'C': 1, 'gamma': 0.1}         -0.058437        -0.092118
21          {'C': 1, 'gamma': 1}         -0.042796        -0.052155
22         {'C': 1, 'gamma': 10}         -0.043097        -0.056414
23        {'C': 1, 'gamma': 100}         -0.046553        -0.070260
24     {'C': 10, 'gamma': 0.001}         -0.063260        -0.106564
25      {'C': 10, 'gamma': 0.01}         -0.045600        -0.079860
26       {'C': 10, 'gamma': 0.1}          0.011945         0.042006
27         {'C': 10, 'gamma': 1}          0.094339         0.254818
28        {'C': 10, 'gamma': 10}          0.052784         0.154655
29       {'C': 10, 'gamma': 100}          0.002360         0.030881
30    {'C': 100, 'gamma': 0.001}         -0.042369        -0.077558
31     {'C': 100, 'gamma': 0.01}          0.098549         0.108446
32      {'C': 100, 'gamma': 0.1}          0.299885         0.524862
33        {'C': 100, 'gamma': 1}          0.205733         0.449199
34       {'C': 100, 'gamma': 10}          0.186232         0.405276
35      {'C': 100, 'gamma': 100}          0.160433         0.327200

In [53]:
print("Best cross-validation accuracy: {:.2f}".format(grid_RadSVM.best_score_))
print("Best parameters: {}".format(grid_RadSVM.best_params_))
print("Train Set Score: {}".format(grid_RadSVM.score(X_reduced_train, Y_train)))
print("Test Set Score: {}".format(grid_RadSVM.score(X_reduced_test, Y_test)))

Best cross-validation accuracy: 0.52
Best parameters: {'C': 100, 'gamma': 0.1}
Train Set Score: 0.2950468960459115
Test Set Score: 0.6915495856779722


In [ ]:
report_table = report_table + [['SVM kernel Rbf', grid_RadSVM.best_params_, grid_RadSVM.score(X_reduced_train, Y_train), grid_RadSVM.score(X_reduced_test, Y_test)]]

### 9 : Support Vector Machine with Kernel

In [55]:
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVR
model_PolySVM = SVR(kernel = 'poly', degree = 2)
param_PolySVM = {'C': [0.001, 0.01, 0.1, 1, 10, 100],'gamma': [0.001, 0.01, 0.1, 1, 10, 100]}
print("Defined Parameters:\n{}".format(param_PolySVM))

Defined Parameters:
{'C': [0.001, 0.01, 0.1, 1, 10, 100], 'gamma': [0.001, 0.01, 0.1, 1, 10, 100]}


In [56]:
report = pd.DataFrame(report_table,columns = ['Model name', 'Model parameter', 'Train accuracy', 'Test accuracy'])
report['Test accuracy']=report['Test accuracy'].apply(lambda x: '%.4f' % x)
report['Train accuracy']=report['Train accuracy'].apply(lambda x: '%.4f' % x)
report

Model name                                    Model parameter  \
0      Linear Regression                                               None   
1         KNN Regression                                              K = 5   
2       Ridge Regression                                        alpha = 0.3   
3       Lasso Regression                                          alpha = 1   
4  Polynomial Regression                                         Degree = 2   
5         SGD Regression  {'learning_rate': 'optimal', 'max_iter': 1000,...   
6             Linear SVM                                        {'C': 1000}   
7         SVM kernel Rbf                           {'C': 100, 'gamma': 0.1}   

  Train accuracy Test accuracy  
0         0.7884        0.7233  
1         0.7730        0.6634  
2         0.8330        0.8061  
3         0.8751        0.7333  
4        -1.1268       -1.1268  
5         0.8753        0.6966  
6         0.8247        0.8100  
7         0.2950        0.6915

### Linear SVM has the highest Test & Train Accuracies

In [57]:
report.iloc[6,:]

Model name          Linear SVM
Model parameter    {'C': 1000}
Train accuracy          0.8247
Test accuracy           0.8100
Name: 6, dtype: object

### Models without PCA  
#### Copied from Project 2

              model	       cross val train	 cross val test	 train r2	test r2
0	K Neighbor Regression	  0.878637	       0.844452	     0.885631	0.872666
1    	Linear Regression	  0.909266	       0.928121	     0.946542	0.930354
2    	 Ridge Regression	  0.913961	       0.926241	     0.938993	0.926606
3    	   SGD Regression	  0.927683	       0.928711	     0.935406	0.929586
4    	 Lasso Regression	  0.878637	       0.844452	     0.885631	0.872666
5	Linear SVM Regression	  0.878632	       0.844464	     0.875612	0.882561
6   Polynomial Regression     0.865917	       0.849450	     0.815620	0.817351

In [59]:
report #Models with PCA

Model name                                    Model parameter  \
0      Linear Regression                                               None   
1         KNN Regression                                              K = 5   
2       Ridge Regression                                        alpha = 0.3   
3       Lasso Regression                                          alpha = 1   
4  Polynomial Regression                                         Degree = 2   
5         SGD Regression  {'learning_rate': 'optimal', 'max_iter': 1000,...   
6             Linear SVM                                        {'C': 1000}   
7         SVM kernel Rbf                           {'C': 100, 'gamma': 0.1}   

  Train accuracy Test accuracy  
0         0.7884        0.7233  
1         0.7730        0.6634  
2         0.8330        0.8061  
3         0.8751        0.7333  
4        -1.1268       -1.1268  
5         0.8753        0.6966  
6         0.8247        0.8100  
7         0.2950        0.6915

### If we compare both table reports, we find that applying PCA hasn't increased the accuracy of the models. It can therefore be concluded that applying dimensionality reduction on our models can be avoided for this specific dataset.

## Deep Learning Networks

In [88]:
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.metrics import r2_score, recall_score, precision_score
from keras.models import Sequential
from keras.layers import Dense

In [64]:
def create_model():
    model = Sequential()
    model.add(Dense(15, input_dim=77, kernel_initializer='normal', activation='relu'))
    model.add(Dense(7, kernel_initializer = 'normal', activation = 'relu'))
    model.add(Dense(1, kernel_initializer='normal'))
    model.compile(optimizer='adam' , loss = 'mse',  metrics = ['mse'])
    return model

In [105]:
model = KerasRegressor(build_fn = create_model, verbose = 0)
param_grid = {'batch_size':[10,20,30,40] , 'epochs':[10, 50, 100]}
grid_search = GridSearchCV(estimator= model, param_grid = param_grid, cv = 5)
grid_search_result = grid_search.fit(X_train, Y_train)

In [106]:
print("Best parameters: {}".format(grid_search.best_params_))
print("Train Set Score: {}".format(grid_search.score(X_train, Y_train)))
print("Test Set Score: {}".format(grid_search.score(X_test, Y_test)))

Best parameters: {'batch_size': 20, 'epochs': 100}
Train Set Score: -68369.72014811198
Test Set Score: -25302.203154296876


In [107]:
model.fit(X_train, Y_train, epochs = grid_search_result.best_params_['epochs'], batch_size =grid_search_result.best_params_['batch_size'] )

In [78]:
model.fit(X_train, Y_train, epochs = 100, batch_size = 10)

In [108]:
y_train_predict = model.predict(X_train)
y_test_predict = model.predict(X_test)

print('Train score: {:.2f}'.format(r2_score(Y_train, y_train_predict)))
print('Test score: {:.2f}'.format(r2_score(Y_test, y_test_predict)))

Train score: 0.81
Test score: 0.77
